src/trajectory.jl

```diff
@@ -227,22 +227,52 @@ samplecand(rng, τ::StaticTrajectory{EndPointTS}, h, z) = step(τ.integrator, h,
 
 ### Multinomial sampling from trajecory
 
-function randcat(rng::AbstractRNG, xs, p)
-    u = rand(rng)
+function randcat(rng::AbstractRNG, p_unorm::AbstractVector{T}) where {T}
+    p = p_unorm ./ sum(p_unorm)
+    u = rand(rng, T)
     cp = zero(eltype(p))
     i = 0
     while cp < u
-        cp += p[i +=1]
+        cp += p[i+=1]
     end
-    return xs[max(i, 1)]
+    return max(i, 1)
+end
+
+function randcat(rng, P_unorm::AbstractMatrix{T}) where {T}
+    P = P_unorm ./ sum(P_unorm; dims=2)
+    u = rand(rng, T, size(P, 1))
+    C = cumsum(P; dims=2)
+    is = convert.(Int, vec(sum(C .< u; dims=2)))
+    return max.(is, 1)
+end
+
+randcat(rng::AbstractRNG, xs, p::AbstractVector) = xs[randcat(rng, p)]
+
+# xs is in the form of Vector{Vector} and has shape [n_steps][]
+function randcat(rng, xs, P_unorm::AbstractMatrix)
+    x = similar(first(xs))
+    is = randcat(rng, P_unorm)
+    foreach(enumerate(is)) do (i_chain, i_step)
+        xi = xs[i_step]
+        x.θ[:,i_chain] = xi.θ[:,i_chain]
+        x.r[:,i_chain] = xi.r[:,i_chain]
+        x.ℓπ.value[i_chain] = xi.ℓπ.value[i_chain]
+        x.ℓπ.gradient[:,i_chain] = xi.ℓπ.gradient[:,i_chain]
+        x.ℓκ.value[i_chain] = xi.ℓκ.value[i_chain]
+        x.ℓκ.gradient[:,i_chain] = xi.ℓκ.gradient[:,i_chain]
+    end
+    return x
 end
 
 function samplecand(rng, τ::StaticTrajectory{MultinomialTS}, h, z)
     zs = step(τ.integrator, h, z, τ.n_steps; res=[z for _ in 1:abs(τ.n_steps)])
     ℓws = -energy.(zs)
+    if eltype(ℓws) <: AbstractVector
+        ℓws = hcat(ℓws...)
+    end
     ℓws = ℓws .- maximum(ℓws)
-    p_unorm = exp.(ℓws)
-    return randcat(rng, zs, p_unorm / sum(p_unorm))
+    prob_unorm = exp.(ℓws)
```

Comment:
Maybe switch to `ℓprob_unorm` here, as the log scale is numerically more stable, particularly when some candidates have a tiny weight while others have a large one.

Reply:
Good point.  Let me change all to log domain.

Reply:
That's true - it might still be good for consistency to always keep things in the log scale to avoid accidental mistreatments. What do you think?

Reply:
Yes I think it would make the code clear. Let's assume `randcat` takes an unnormalised log probability then.

Reply:
If so I guess we should renmae `randcat` to something else that indicates it takes probabilities in log space.

Another question is do we want to keep probability in log space even within `randcat`? If so it would requires change all accumation of probabities to take use of `logaddexp`, which I assume would be slower than simply doing addition or `cumsum`.

Reply:
It's ok to work with ordinary probabilities in `randcat` I think. We can revisit this later if it causes numerical instabilities.

Reply:
Cool. I renamed the function which takes log probabilities as `randcatlog` to indicate the input is in log domain.